### PR TITLE
GCC: Add bootstrap conflict for M1

### LINKED ISSUE #31179: Installation issue: gcc not building on macOS Monterey with M1
### Steps to reproduce the issue

Gcc is not building for me on Monetary on a new M1 MacBook Pro.
```console
$ spack install gcc@11.3.0
```

@alalazo @michaelkuhn 

### Error message

<details>
There seems to be a bootstrapping issue that needs to be ported.
<summary>Error message</summary><pre>
     551051    rm -f stage_current
     551052    Comparing stages 2 and 3
     551053    Bootstrap comparison failure!
     551054    aarch64-apple-darwin21.5.0/libstdc++-v3/src/.libs/libstdc++.6.dylib-master.o differs
  >> 551055    make[2]: *** [compare] Error 1
  >> 551056    make[1]: *** [stage3-bubble] Error 2
  >> 551057    make: *** [all] Error 2
</pre></details>

### Information on your system

```console
spack debug report
* **Spack:** 0.19.0.dev0 (bf990bc8ec2d6c51f7f40bc6b0f7fb3bba897a96)
* **Python:** 3.9.13
* **Platform:** darwin-monterey-m1
* **Concretizer:** clingo
```

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/8928055/spack-build-env.txt)
[spack-build-out.txt.zip](https://github.com/spack/spack/files/8928061/spack-build-out.txt.zip)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by psakievich**
Closes #31179

@sethrj @alalazo @michaelkuhn

### CODE REVIEW

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 1]
```diff
@@ -297,6 +297,7 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
         patch('https://raw.githubusercontent.com/Homebrew/formula-patches/22dec3fc/gcc/gcc-11.3.0-arm.diff',
               sha256='e02006b7ec917cc1390645d95735a6a866caed0dfe506d5bef742f7862cab218',
               when='@11.3.0 target=aarch64:')
+        conflicts('+bootstrap', when='@11.3.0 target=aarch64')
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/gcc/package.py]:
```suggestion
        conflicts('+bootstrap', when='@11.3.0 target=aarch64:')
```
Open ended range needed to cover M1 and other specific architectures

**Comment by adamjstewart** [comment]
`gcc@11.3.0+bootstrap` installed for me without issue on macOS 12.4 (Apple M1 Pro).

EDIT: this is with Apple Clang 13.1.6.

**Comment by sethrj** [comment]
Maybe it's a clang version difference
? What are y'all's toolchain versions?

**Comment by psakievich** [comment]
Strange. I'm using the same thing.
```console
$ spack compilers
-- apple-clang monterey-aarch64 ---------------------------------
apple-clang@13.1.6

$ gcc --version
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: arm64-apple-darwin21.5.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bi
```

**Comment by sethrj** [comment]
It's probably the new Xcode and Apple clang...

**Comment by psakievich** [comment]
Most likely. I'm just setting this up and applied all the developer updates. @adamjstewart have you updated developer tools recently?

**Comment by adamjstewart** [comment]
I'm on the latest version of everything, but I think we have the same version of clang:
```console
$ clang --version
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: arm64-apple-darwin21.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Review by William-Mou — COMMENTED** [review]
It's work for me!
* **Spack:** 0.19.0.dev0 (7741bfb7d1628d8f2e17284e34cbb8d4a69f3553)
* **Python:** 3.9.13
* **Platform:** darwin-monterey-m1
* **Concretizer:** clingo

**Comment by psakievich** [comment]
Thoughts on merging this? @adamjstewart I tried building gcc again this morning from a fresh spack repo and I'm still seeing  it fail with bootstrapping enabled.

Here is the DAG I'm working off if you want to compare:
```console
spack spec -Il gcc%apple-clang
Input spec
--------------------------------
 -   gcc%apple-clang

Concretized
--------------------------------
 -   fl7syxm  gcc@11.3.0%apple-clang@13.1.6~binutils+bootstrap~graphite~nvptx~piclibs~strip build_type=RelWithDebInfo languages=c,c++,fortran patches=e02006b arch=darwin-monterey-m1
[+]  mbrctz7      ^diffutils@3.8%apple-clang@13.1.6 arch=darwin-monterey-m1
[+]  6ge5mev          ^gnuconfig@2021-08-14%apple-clang@13.1.6 arch=darwin-monterey-m1
[+]  fh6dfpn          ^libiconv@1.16%apple-clang@13.1.6 libs=shared,static arch=darwin-monterey-m1
[+]  tjeatah      ^gawk@5.1.1%apple-clang@13.1.6~nls arch=darwin-monterey-m1
[+]  twbspbd          ^gmp@6.2.1%apple-clang@13.1.6 libs=shared,static arch=darwin-monterey-m1
[+]  2fykb5r              ^autoconf@2.69%apple-clang@13.1.6 patches=35c4492,7793209,a49dd5b arch=darwin-monterey-m1
[+]  y4neus6                  ^m4@1.4.19%apple-clang@13.1.6+sigsegv patches=9dc5fbd,bfdffa7 arch=darwin-monterey-m1
[+]  t5akhjv                      ^libsigsegv@2.13%apple-clang@13.1.6 arch=darwin-monterey-m1
[+]  p5gcujt                  ^perl@5.34.1%apple-clang@13.1.6+cpanm+shared+threads arch=darwin-monterey-m1
[+]  e5dqqjq                      ^berkeley-db@18.1.40%apple-clang@13.1.6+cxx~docs+stl patches=b231fcc arch=darwin-monterey-m1
[+]  rjsj6wi                      ^bzip2@1.0.8%apple-clang@13.1.6~debug~pic+shared arch=darwin-monterey-m1
[+]  gc2prnk                      ^gdbm@1.19%apple-clang@13.1.6 arch=darwin-monterey-m1
[+]  m43uguo                          ^readline@8.1%apple-clang@13.1.6 arch=darwin-monterey-m1
[+]  muw42gz                              ^ncurses@6.2%apple-clang@13.1.6~symlinks+termlib abi=none arch=darwin-monterey-m1
[+]  ett7ygi                                  ^pkgconf@1.8.0%apple-clang@13.1.6 arch=darwin-monterey-m1
[+]  ke4alug                      ^zlib@1.2.12%apple-clang@13.1.6+optimize+pic+shared patches=0d38234 arch=darwin-monterey-m1
[+]  oiy6xiy              ^automake@1.16.5%apple-clang@13.1.6 arch=darwin-monterey-m1
[+]  3zf2uh2              ^libtool@2.4.7%apple-clang@13.1.6 arch=darwin-monterey-m1
[+]  qefruzv          ^mpfr@4.1.0%apple-clang@13.1.6 libs=shared,static arch=darwin-monterey-m1
[+]  a2tiepv              ^autoconf-archive@2022.02.11%apple-clang@13.1.6 patches=130cd48 arch=darwin-monterey-m1
[+]  7ldfnbz              ^texinfo@6.5%apple-clang@13.1.6 patches=12f6edb,1732115 arch=darwin-monterey-m1
[+]  gerdw6o      ^mpc@1.2.1%apple-clang@13.1.6 libs=shared,static arch=darwin-monterey-m1
[+]  r7rfcbo      ^zstd@1.5.2%apple-clang@13.1.6~programs libs=shared,static arch=darwin-monterey-m1
```

**Comment by adamjstewart** [comment]
> `gcc@11.3.0+bootstrap` installed for me without issue on macOS 12.4 (Apple M1 Pro).

I stand corrected. I just tried again and found that I also encounter the same issue with `gcc+bootstrap`. Using `gcc~bootstrap` allows me to build the package. So this PR LGTM.

**Review by adamjstewart — APPROVED** [review]

**Comment by sethrj** [comment]
Thanks @adamjstewart !

**Review by sethrj — APPROVED** [review]

**Pull request merged**